### PR TITLE
core: msg: update detail section on IPC

### PR DESCRIPTION
**Pull request by authmillenon**
Merges IPC section of old doxygen mainpage and detailed description and enhances it with some introductory text and examples.

Since I have uncrustify run as a pre-commit hook I included some style fixes too in an extra commit.

### CODE REVIEW

**Comment by authmillenon** [comment]
Arch users be advices that there is a bug in the most current doxygen version regarding the fenced code block syntax: https://bugzilla.gnome.org/show_bug.cgi?id=759177

File: core/include/msg.h [hunk 1]
```diff
@@ -11,14 +11,120 @@
...
+ * user. The type helps the receiver to multiplex different message types and
+ * should be set to a system-wide unique value. The content can either be
+ * provided as a 32-bit integer or a pointer.
+ *
+ * Synchronous IPC
+ * ===============
+ * To send a message synchronous to a thread one uses the function
+ * @ref msg_send_receive(). It sends a message and blocks the sending thread
+ * until a response is received. The receiving thread uses the function
+ * @ref msg_receive() to receive this message and must must reply to it with
```
Comment (kaspar030) [inline core/include/msg.h]: must must

File: core/include/msg.h [hunk 2]
```diff
@@ -11,14 +11,120 @@
...
- * with a full message queue behaves like in synchronous mode.
+ * Messages
+ * ========
+ * IPC messages consist of a sender PID, a type, and some content. The sender
+ * PID will be set by the IPC internally and is not required to be set by the
+ * user. The type helps the receiver to multiplex different message types and
+ * should be set to a system-wide unique value. The content can either be
+ * provided as a 32-bit integer or a pointer.
+ *
+ * Synchronous IPC
```
Comment (kaspar030) [inline core/include/msg.h]: This is misleading. Synchronous means "no message queue".

File: core/include/msg.h [hunk 3]
```diff
@@ -11,14 +11,120 @@
...
+ * #include <inttypes.h>
+ * #include <stdio.h>
+ *
+ * #include "msg.h"
+ * #include "thread.h"
+ *
+ * #define RCV_QUEUE_SIZE  (8)
+ *
+ * kernel_pid_t rcv_pid;
+ * char rcv_stack[THREAD_STACKSIZE_DEFAULT + THREAD_EXTRA_STACKSIZE_PRINTF];
```
Comment (kaspar030) [inline core/include/msg.h]: static

File: core/include/msg.h [hunk 4]
```diff
@@ -11,14 +11,120 @@
...
+ *         }
+ *         msg.content.value++;
+ *     }
+ *     return 0;
+ * }
+ * ~~~~~~~~~~~~~~~~~~~~~~~~
+ *
+ * Time & messages
+ * ===============
+ * Timeing out a receive or sending messages at a certain time is out of scope
```
Comment (kaspar030) [inline core/include/msg.h]: timing

**Comment by kaspar030** [comment]
thx for taking the time to create examples!

**Comment by authmillenon** [comment]
What example would I set if I would not give examples ;P

File: core/include/msg.h (hunk 2 above)
Comment (authmillenon) [inline core/include/msg.h]: That's news to me. Then I need to restructure this a little. The current
structure assumes the described behavior is the synchronous IPC.

Comment (kaspar030) [inline core/include/msg.h]: If you look at the diff to the left, it describes messging using message queues as asynchronous.

Comment (authmillenon) [inline core/include/msg.h]: I've read it as statement that message queues are required for asynchronous IPC, but optional for what I understood as synchronous. Hopefully, my update will lift this confusion for future (and current developers).

**Comment by authmillenon** [comment]
addressed comments

File: core/include/msg.h [hunk 5]
```diff
@@ -11,14 +11,141 @@
...
+ * received with the non-blocking variant and vice-versa.
+ *
+ * Blocking IPC
+ * ------------
+ * For the blocking variant use @ref msg_send() or @ref msg_receive()
+ * respectively.
+ *
+ * Additionally, one can use @ref msg_send_receive() to simultaneously block
+ * the sending thread and expect a response response from the receiving thread.
+ * The receiving thread must use @ref msg_reply() to reply to the message of
```
Comment (kaspar030) [inline core/include/msg.h]: "In this case, ..."

File: core/include/msg.h [hunk 6]
```diff
@@ -11,14 +11,141 @@
...
+ * messages send this way will be dropped.
+ *
+ * As an example you can use the example on asynchronous IPC below without the
+ * queue.
+ *
+ * Synchronous vs Asynchronous
+ * ===========================
+ * RIOT's IPC supports both synchronous and asynchronous IPC.
+ *
+ * Synchronous IPC is the default mode and one does not need to anything to use
```
Comment (kaspar030) [inline core/include/msg.h]: do

File: core/include/msg.h [hunk 7]
```diff
@@ -11,14 +11,141 @@
...
+ * it.
+ *
+ * To use asynchronous IPC one needs to initialize a message queue using
+ * @ref msg_init_queue(). Messages sent to a thread with a message queue that
+ * isn't full are never dropped and the sending never blocks, even when using
+ * @ref msg_send(). If the queue is full and the sending thread of a higher
```
Comment (kaspar030) [inline core/include/msg.h]: "of a higher" -> "has a higher priority" ?

**Comment by kaspar030** [comment]
minor comments, but from my side, you can squash.

File: core/include/msg.h (hunk 6 above)
Comment (authmillenon) [inline core/include/msg.h]: no.

Comment (authmillenon) [inline core/include/msg.h]: ("one" is 3rd person singular, so you need to use "does").

**Comment by authmillenon** [comment]
Addressed comments and squashed immediately.

File: core/include/msg.h (hunk 6 above)
Comment (kaspar030) [inline core/include/msg.h]: sorry, I meant "...one does not need to *do* anything to use ..."

Comment (authmillenon) [inline core/include/msg.h]: Fixed and squashed

**Comment by authmillenon** [comment]
Does this need to ACKs due to it being a core change?

**Comment by kaspar030** [comment]
nah. The two line code change doesn't chang anything semantically, so one ACK is OK.

File: core/include/msg.h [hunk 8]
```diff
@@ -11,14 +11,141 @@
...
+ *         printf("Result: %" PRIu32 "\n", msg_resp.content.value);
+ *     }
+ *     return 0;
+ * }
+ * ~~~~~~~~~~~~~~~~~~~~~~~~
+ *
+ * Non-blocking IPC
+ * ----------------
+ * For the non-blocking variant use @ref msg_try_send() or
+ * @ref msg_try_receive() respectively. If a message is send in synchronous
```
Comment (kaspar030) [inline core/include/msg.h]: "is sent"

File: core/include/msg.h [hunk 9]
```diff
@@ -11,14 +11,141 @@
...
+ * mode or the message queue (see below) of the receiving thread is full
+ * messages send this way will be dropped.
```
Comment (kaspar030) [inline core/include/msg.h]: "messages sent"

**Comment by kaspar030** [comment]
found two more typos... fix if you feel like... IMHO we should skip travis on this (all but the static tests that check the docs).

File: core/include/msg.h [hunk 10]
```diff
@@ -11,14 +11,141 @@
...
+ * used combined: A message send while blocking the sender thread can be
+ * received with the non-blocking variant and vice-versa.
+ *
+ * Blocking IPC
+ * ------------
+ * For the blocking variant use @ref msg_send() or @ref msg_receive()
+ * respectively.
+ *
+ * Additionally, one can use @ref msg_send_receive() to simultaneously block
+ * the sending thread and expect a response response from the receiving thread.
```
Comment (PeterKietzmann) [inline core/include/msg.h]: response response

File: core/include/msg.h [hunk 11]
```diff
@@ -11,14 +11,141 @@
...
+ * ~~~~~~~~~~~~~~~~~~~~~~~~
+ *
+ * Non-blocking IPC
+ * ----------------
+ * For the non-blocking variant use @ref msg_try_send() or
+ * @ref msg_try_receive() respectively. If a message is send in synchronous
+ * mode or the message queue (see below) of the receiving thread is full
+ * messages send this way will be dropped.
+ *
+ * As an example you can use the example on asynchronous IPC below without the
```
Comment (PeterKietzmann) [inline core/include/msg.h]: Maybe reword one "example"

File: core/include/msg.h [hunk 12]
```diff
@@ -11,14 +11,141 @@
...
+ * queue.
+ *
+ * Synchronous vs Asynchronous
+ * ===========================
+ * RIOT's IPC supports both synchronous and asynchronous IPC.
+ *
+ * Synchronous IPC is the default mode and one does not need to do anything to
```
Comment (PeterKietzmann) [inline core/include/msg.h]: So you actively decided against subsections for synchronous and asynchronous IPC here (like you did for blocking/non-blocking)? It's not a strong opinion but I liked it the way you did above.

Comment (PeterKietzmann) [inline core/include/msg.h]: > one does not need to do anything to use it

I stumbled upon that. What would be the opposite? Maybe it's more precise to state **what** you **don't** have to do.

File: core/include/msg.h [hunk 13]
```diff
@@ -11,14 +11,141 @@
...
+ *         if (msg_try_send(&msg, rcv_pid) == 0) {
+ *             printf("Receiver queue full.\n");
+ *         }
+ *         msg.content.value++;
+ *     }
+ *     return 0;
+ * }
+ * ~~~~~~~~~~~~~~~~~~~~~~~~
+ *
+ * Time & messages
```
Comment (PeterKietzmann) [inline core/include/msg.h]: How about "Tim**ing** & messages"?

**Comment by PeterKietzmann** [comment]
Thanks Martine for taking the initiative! No need to discuss on my proposed changes, adapt what you take as reasonable.

File: core/include/msg.h (hunk 12 above)
Comment (authmillenon) [inline core/include/msg.h]: I decided against subsections, because synchronous IPC was just one sentence, but I can use it anyway.

**Comment by authmillenon** [comment]
Addressed last batch of comments.

File: core/include/msg.h [hunk 14]
```diff
@@ -11,14 +11,147 @@
...
+ * ~~~~~~~~~~~~~~~~~~~~~~~~
+ *
+ * Non-blocking IPC
+ * ----------------
+ * For the non-blocking variant use @ref msg_try_send() or
+ * @ref msg_try_receive() respectively. If a message is sent in synchronous
+ * mode or the message queue (see below) of the receiving thread is full
+ * messages sent this way will be dropped.
+ *
+ * You can use the example on asynchronous IPC below but without the queue for
```
Comment (PeterKietzmann) [inline core/include/msg.h]: "for" too much I guess

File: core/include/msg.h [hunk 15]
```diff
@@ -11,14 +11,147 @@
...
+ *
+ * Synchronous vs Asynchronous
+ * ===========================
+ * RIOT's IPC supports both synchronous and asynchronous IPC.
+ *
+ * Synchronous IPC
+ * ---------------
+ * Synchronous IPC is the default mode i.e. is active when the receiving thread
+ * has no message queue initialized. Messages that can't be delivered when
+ * sending non-blocking (because the receiver already received a message) or
```
Comment (PeterKietzmann) [inline core/include/msg.h]: Easier to read: ...or **which** are sent ...

**Comment by PeterKietzmann** [comment]
@authmillenon sorry being so picky, but now that you're on it ;-) ! IMO it's fine to squash

**Comment by authmillenon** [comment]
Addressed comments and squashed

**Comment by PeterKietzmann** [comment]
ACK from my side. @kaspar030 please have a look and merge then!!!

File: core/include/msg.h [hunk 16]
```diff
@@ -148,7 +281,7 @@ int msg_send_int(msg_t *m, kernel_pid_t target_pid);
  */
 static inline int msg_sent_by_int(const msg_t *m)
 {
-    return (m->sender_pid == KERNEL_PID_ISR);
+    return m->sender_pid == KERNEL_PID_ISR;
```
Comment (OlegHahm) [inline core/include/msg.h]: Why did you remove these parens?

Comment (kaspar030) [inline core/include/msg.h]: It's in the piggybacked uncrustify commit.

Comment (OlegHahm) [inline core/include/msg.h]: I guess we have to fix some settings there.

Comment (authmillenon) [inline core/include/msg.h]: It's one expression, so as far as I understand your parenthesis rules this should be right...

Comment (authmillenon) [inline core/include/msg.h]: (for e.g. two expressions the parenthesis would be kept around each)

Comment (authmillenon) [inline core/include/msg.h]: I think it's caused by `mod_paren_on_return = remove`, but setting this to ignore or even add would mean that e.g. `return (1);` would be correct, too.

Comment (OlegHahm) [inline core/include/msg.h]: "My" rule is basically always to have parens around any condition. Especially, I don't see any reason to remove them.

Comment (OlegHahm) [inline core/include/msg.h]: IMO `ignore` sounds good.

Comment (authmillenon) [inline core/include/msg.h]: Changed and squashed immediately.

Comment (authmillenon) [inline core/include/msg.h]: https://github.com/RIOT-OS/RIOT/pull/4481

Comment (OlegHahm) [inline core/include/msg.h]: thx

**Comment by authmillenon** [comment]
I'm confused. @PeterKietzmann did you ACK or do you want @kaspar030 to ACK?

**Comment by kaspar030** [comment]
IMHO its fine. Go!

**Comment by authmillenon** [comment]
Release version in #4496
